### PR TITLE
Sealed typed arrays should report themselves frozen only if empty

### LINKED ISSUE #5624: Object.seal(TypedArray) should not freeze object as well
Hi everyone,
there is an inconsistency in Chakra if we seal a TypedArray the object is frozen as well. According to [specs](https://tc39.github.io/ecma262/#sec-setintegritylevel), Seal level should set only properties configurable to false.

version: 1.10.2
OS: Ubuntu 16.04 x64

steps to reproduce:
```js
let ta = new Int32Array(10);
Object.seal(ta);
if (Object.isExtensible(ta) !== false) throw new Error('Object.isExtensible should be false')
if (Object.isSealed(ta) !== true) throw new Error('Object.isSealed should be true')
if (Object.isFrozen(ta) !== false) throw new Error('Object.isFrozen should be false')
```
Actual results:
Error: Object.isFrozen should be false

Expected results:
pass without failures

SpiderMonkey works as expected, V8 fails because it cannot seal this object and JSC fails due non-enumerable/non-writable property on a typed array.

cinfuzz

### PR DESCRIPTION
**Pull request by irinayat-MS**
Fixes #5624

Rewritten and expanded tests for seal/freeze. 
Fixed the bug.

### CODE REVIEW

File: test/es5/freeze.js [hunk 1]
```diff
@@ -3,47 +3,117 @@
...
+
+          // cannot add new properties
+          a.y = 17;
+          assert.isFalse(a.hasOwnProperty('y'));
+          assert.throws(function () { 'use strict'; a.y = 17; }, TypeError, 
+            "Should throw on creating a new property in frozen object in strict mode", 
+            "Cannot create property for a non-extensible object");
+
+          // cannot delete properties
+          delete a.x;
```
Comment (jackhorton) [inline test/es5/freeze.js]: may want to assert that this returns false.

Comment (irinayat-MS) [reply test/es5/freeze.js]: done

File: test/es5/freeze.js [hunk 2]
```diff
@@ -3,47 +3,117 @@
...
+    name: "Add, delete, modify indexed elements of a typed array after freezing",
+    body: function () {
+        let a = new Int8Array(1);
+        a[0] = 42;
+
+        Object.freeze(a); // should it throw?
+
+        assert.isFalse(Object.isExtensible(a));
+        assert.isTrue(Object.isSealed(a));
+        assert.isTrue(Object.isFrozen(a)); // should it return false?
```
[the remark this answers is not in the document]
Comment (irinayat-MS) [reply test/es5/freeze.js]: Explicitly freezing an object sets IsFrozenOnceFlag flag so it doesn't invoke IsObjectArrayFrozen (even though it still allows to update the elements). I'll add the flag check to IsObjectArrayFrozen for consistency.

---
In reply to: [213745165](https://github.com/Microsoft/ChakraCore/pull/5652#discussion_r213745165) [](ancestors = 213745165)

File: test/es5/seal.js [hunk 3]
```diff
@@ -3,42 +3,175 @@
...
+  },
+  {
+    name: "Add, delete, modify indexed elements of a typed array after sealing",
+    body: function () {
+        let a = new Int8Array(1);
+        a[0] = 42;
+
+        Object.seal(a);
+        assert.isFalse(Object.isExtensible(a));
+        assert.isTrue(Object.isSealed(a));
```
[the remark this answers is not in the document]
Comment (irinayat-MS) [reply test/es5/seal.js]: Yes, I think this is the case when we are more spec compliant than v8 :)

---
In reply to: [213746309](https://github.com/Microsoft/ChakraCore/pull/5652#discussion_r213746309) [](ancestors = 213746309)

File: test/es5/freeze.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (irinayat-MS) [reply test/es5/freeze.js]: Actually, the flag is on the handler, not the type itself so cannot check it directly... Will have to rely on the existing logic that checks the flag first before re-dispatching for more thorough (perf-heavy) checks.

---
In reply to: [213772620](https://github.com/Microsoft/ChakraCore/pull/5652#discussion_r213772620) [](ancestors = 213772620,213745165)

File: test/es5/freeze.js [hunk 4]
```diff
@@ -3,47 +3,117 @@
...
+        assert.isFalse(descr.configurable);
+        assert.isFalse(descr.writable);
+        a[0] = 17;
+        assert.areEqual(42, a[0]);
+        assert.throws(function () { 'use strict'; a[0] = 17; }, TypeError, 
+          "Should throw on creating a new property in frozen object in strict mode", 
+          "Assignment to read-only properties is not allowed in strict mode");
+
+        // the special 'length' property also cannot be modified
+        let descr_len = Object.getOwnPropertyDescriptor(a, '0');
```
[the remark this answers is not in the document]
Comment (irinayat-MS) [reply test/es5/freeze.js]: Thanks!

---
In reply to: [213744540](https://github.com/Microsoft/ChakraCore/pull/5652#discussion_r213744540) [](ancestors = 213744540)

**Pull request merged**